### PR TITLE
OIDC: allow configuring the request timeout

### PR DESCRIPTION
**Pull request by fflorent**
## Context

I take up this PR: #1072 (open by @atropos112)

In some case, the identity provider may take time before responding to requests sent by the OIDC client, leading to difficulties or even impossibility to start the server or to login.

For example, if the issuer discovery takes time, you may fail to start the server with this error:
```
RPError: outgoing request timed out after 3500ms
    at /home/florentpro/projects/grist-core/node_modules/openid-client/lib/helpers/request.js:137:13
    at async Issuer.discover (/home/florentpro/projects/grist-core/node_modules/openid-client/lib/issuer.js:171:22)
    at async OIDCConfig._initClient (/home/florentpro/projects/grist-core/_build/app/server/lib/OIDCConfig.js:246:24)
    at async OIDCConfig.initOIDC (/home/florentpro/projects/grist-core/_build/app/server/lib/OIDCConfig.js:155:9)
    at async OIDCConfig.build (/home/florentpro/projects/grist-core/_build/app/server/lib/OIDCConfig.js:104:9)
    at async Object.getMiddleware (/home/florentpro/projects/grist-core/_build/app/server/lib/OIDCConfig.js:330:28)
    at async FlexServer.addLoginMiddleware (/home/florentpro/projects/grist-core/_build/app/server/lib/FlexServer.js:1083:33)
    at async main (/home/florentpro/projects/grist-core/_build/app/server/mergedServerMain.js:86:5)
    at async main (/home/florentpro/projects/grist-core/_build/stubs/app/server/server.js:143:20)
```

## Proposed solution

Introduce the `GRIST_OIDC_SP_HTTP_TIMEOUT` env variable so the user may set a greater value than the default 3500ms, or even [set it to 0 to remove any timeout](https://nodejs.org/api/https.html#servertimeout).

## How to test it

I use this utility to add latency to the requests, so the openid-client requests may timeout: https://github.com/sitespeedio/throttle

Prerequisites:
 - Install throttle (introduced with a link above): `npm i -g @sitespeed.io/throttle`
 - Install `tc`. On Debian-family Linux distros, you must install iproute2 for that: `sudo apt install iproute2`.

STR:
 - Setup a Keycloak environment and use it
 - Configure Grist to use Keycloak as IdP: https://github.com/gristlabs/grist-help/blob/master/help/en/docs/install/oidc.md#example-keycloak 
 - Use throttle to add the latency. I run it this way: `throttle --localhost --up 9000 --down 9000 --rtt 5000` (the RTT is what matters here)
 - Then:
   - to reproduce the issue: `yarn start`, the server should fail to start with a timeout in the logs (see the `Context` section above)
   - or `GRIST_OIDC_SP_HTTP_TIMEOUT=30000 yarn start` to check that the problem is solved
 - Once you have finished testing, you may stop the latency: `throttle --localhost stop`

## Related issues

I also pave the way for #942

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help

### CODE REVIEW

**Review by hexaltation — APPROVED** [review]
LGTM

File: app/server/lib/OIDCConfig.ts [hunk 1]
```diff
@@ -173,6 +178,9 @@ export class OIDCConfig {
     this._protectionManager = new ProtectionsManager(enabledProtections);
 
     this._redirectUrl = new URL(CALLBACK_URL, spHost).href;
+    custom.setHttpOptionsDefaults({
+      ...(httpTimeout !== undefined ? {timeout: httpTimeout} : {}),
```
Comment (dsagal) [inline app/server/lib/OIDCConfig.ts:182]: Why not more simply:
```
if (httpTimeout !== undefined) {
  custom.setHttpOptionsDefaults({timeout: httpTimeout});
}
```

**Review by dsagal — COMMENTED** [review]
Looks good, just one cosmetic question.

File: app/server/lib/OIDCConfig.ts (hunk 1 above)
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:182]: Hi @dsagal!

For slightly convenient reasons:
 1. I don't have to care whether the function should have been called or not in the test, it should if no exception has been raised: https://github.com/gristlabs/grist-core/pull/1177/files#diff-0467f3ff93b79307ee0beb1d72eea81866d60efa1efcfb8937d7154117e895f4R238
 2. I think of passing later more options to fix #942 

If these are bad reasons to you, I may apply the change you request, I don't have much strong opinions.

**Review by dsagal — APPROVED** [review]

File: app/server/lib/OIDCConfig.ts (hunk 1 above)
Comment (fflorent) [reply app/server/lib/OIDCConfig.ts:182]: Yes, I agree with what you tell (we miss integration tests for this part :grimacing:).
I prefer not taking any assumption (for now and the future) and keep this part as it is.

**Comment by fflorent** [comment]
Thanks @dsagal for your review!

I don't have the rights to merge. If there is no other change you expect from me, could you do that for me if that's OK for you?

**Review by paulfitz — APPROVED** [review]

**Pull request merged**